lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -1034,6 +1034,40 @@ FIRRTLType firrtl::getReductionResult(FIRRTLType input) {
 // Other Operations
 //===----------------------------------------------------------------------===//
 
+static LogicalResult verifyBitsPrimOp(BitsPrimOp bits) {
+  uint32_t hi = bits.hi(), lo = bits.lo();
+
+  // High must be >= low.
+  if (hi < lo) {
+    bits.emitError()
+        << "high must be equal or greater than low, but got high = " << hi
+        << ", low = " << lo;
+    return failure();
+  }
```

Comment:
One thing that I do _not_ know how to untangle the fact that [some of these checks are already done in `getResultType`](https://github.com/llvm/circt/blob/0a1d7ab769a9dd843bd99fe9cbfeafcdbcb4755f/lib/Dialect/FIRRTL/Ops.cpp#L1041), albeit with terrible error reporting relative to this PR.

Concretely, even with this PR, a parsed FIRRTL circuit that should get a nice error, gets a poor one:

```bash
cat Foo.fir
# circuit Foo:
#   module Foo:
#     input a: UInt<8>
#     output b: UInt<4>
#     b <= bits(a, 4, 7)

firtool Foo.fir
# Foo.fir:5:10: error: invalid input types for 'bits': '!firrtl.uint<8>'
#     b <= bits(a, 4, 7)
```

Conversely working on this at the MLIR level gets the excellent diagnostic that this PR adds:

```bash
cat Foo.mlir
# firrtl.circuit "Foo" {
#   firrtl.module @Foo(%a : !firrtl.uint<8>) {
#     %b = firrtl.bits %a 4 to 7 : (!firrtl.uint<8>) -> !firrtl.uint<4>
#   }
# }

firtool Foo.mlir
# Foo.mlir:3:10: error: high must be equal or greater than low, but got high = 4, low = 7
#     %b = firrtl.bits %a 4 to 7 : (!firrtl.uint<8>) -> !firrtl.uint<4>
#          ^
# Foo.mlir:3:10: note: see current operation: %0 = "firrtl.bits"(%a) {hi = 4 : i32, lo = 7 : i32} : (!firrtl.uint<8>) -> !firrtl.uint<4>
```

@lattner: Do you have an idea of how this should be handled?

Reply:
It would be better to catch this in the verifier.  The FIRRTL parser uses `getResultType()`, but other things would generate new nodes directly.  As you mention, such a change also improves the quality of the output.

I think that one way to do this is to have `getResultType()` return something generic like `sint` on errors.

Reply:
Sounds good to me. I had thought about changing `getResultType()` to return what the user meant, i.e., one plus the absolute value of the difference, and rely on the verification here to do the check.

Reply:
I'd work on similar ops like `tail, head, ..` and as you know, their verification is almost equal to what `getResultType()` is doing(i.e the calculation of the result width). There will be many code duplication so I feel it's better to reuse or integrate `getResultType` to the verifier but I'm not sure which is better. What do you think about it?

Reply:
How about changing getResultType to take an argument used to report errors, so it can be called from the parser as well as from the verifier logic?  That would make the parser nicer than just failing with "invalid types" and elimiante the redundancy

Reply:
thank you for working on this @uenoku , this is a great improvement!

Reply:
Sounds good to me. I believe it is sufficient to add `mlir::InFlightDiagnosic` to an argument. I'll merge this PR for now and modify them in the later PR.

Reply:
I think you can just pass a location in.  This will allow reporting errors with these global functions in mlir/IR/Diagnostics.h:

```
/// Utility method to emit an error message using this location.
InFlightDiagnostic emitError(Location loc);
InFlightDiagnostic emitError(Location loc, const Twine &message);
```

Reply:
It makes sense! Thank you!